Fix #3488: Domain layer mechanism to resume explorations.

MaskedCarrot opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation

Fix #3488 
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### Espresso Tests
#### ExplorationActivityTest
![Screenshot from 2021-08-20 03-04-28](https://user-images.githubusercontent.com/53645584/130150246-71e9d4f0-6cee-480e-b229-f741ec6f30fb.png)

#### StateFragmentTest (only the new test added)
![Screenshot from 2021-08-20 03-24-41](https://user-images.githubusercontent.com/53645584/130150290-3710199a-b3c8-4118-9a5a-e272d729f7ff.png)

Commit:
moved hint handler to domain layer

yashraj-01 committed:
fixed app layer espresso tests

yashraj-01 committed:
fix app layer robolectric tests

yashraj-01 committed:
fixed domain layer tests

yashraj-01 committed:
Added annotations to test exemptions

yashraj-01 committed:
proto lint fix

Commit:
fixed hint handler for training sessions

Commit:
nit and removed test excemptions

Commit:
added hint tests for config change

Commit:
fixed test file exemptions

Commit:
fixed failing test

yashraj-01 committed:
made HintHandler injectable

yashraj-01 committed:
Merge remote-tracking branch 'upstream/move-hint-handler-to-domain' into move-hint-handler-to-domain

yashraj-01 committed:
fixed ktlint error

yashraj-01 committed:
Added tests for hint handler

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

Commit:
nits

Commit:
Merge remote-tracking branch 'origin/move-hint-handler-to-domain' into move-hint-handler-to-domain

Commit:
fixed failing build

Commit:
fixed failing build

Commit:
fix build

Commit:
fixed imports

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

Commit:
completed mechansim to resume exploration

BenHenning commented:
Will need to take a look at this tomorrow.

aggarwalpulkit596 commented on app/src/main/java/org/oppia/android/app/player/state/testing/StateFragmentTestActivityPresenter.kt:
```diff
@@ -98,7 +99,8 @@ class StateFragmentTestActivityPresenter @Inject constructor(
       explorationId,
-      shouldSavePartialProgress = false
+      shouldSavePartialProgress = false,
+      explorationCheckpoint = ExplorationCheckpoint.getDefaultInstance()
```
Since we are passing this to every function call and the value is the same too, shouldn't we make it a default argument in the function definition itself? 
@BenHenning @rt4914 Your thoughts?

MaskedCarrot replied:
This is a temporary measure, it is changed in the following PRs ( when we add update the app layer mechanism to start exploration.)

BenHenning replied:
I think I lean toward this being explicit given that one "happy path" is passing something non-default in.

aggarwalpulkit596 commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -720,4 +728,160 @@ class ExplorationProgressController @Inject constructor(
       lastPlayedTimestamp
     )
   }
+
```
Kdocs missing

MaskedCarrot replied:
done

BenHenning replied:
FWIW, KDocs aren't required for private members (per https://developer.android.com/kotlin/style-guide#usage), but they're definitely fine to add.

Commit:
nits and improved testing

Commit:
updated exploration.proto

Commit:
removed progress controller from kdoc exemptions

Commit:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

Commit:
added kdoc

Commit:
fixed failing build

Commit:
Merge remote-tracking branch 'origin/develop' into move-hint-handler-to-domain

Commit:
fix failing test

Commit:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

Commit:
fix failing build

anandwana001 commented:
Assign me back once the UI part gets reviewed, to prevent multiple reviews for test cases. 
Also add screenshot of espresso test result for app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
+   * Updates the hint state in [ExplorationProgrses] with the one created from the saved checkpoint
+   * or the default instance of hint state if exploration is not being resumed.
+   */
+  private fun loadHintState(progress: ExplorationProgress) {
+    progress.hintState =
+      if (progress.currentExplorationCheckpoint == ExplorationCheckpoint.getDefaultInstance()) {
+        HintState.getDefaultInstance()
+      } else {
+        HintState.newBuilder().apply {
+          helpIndex = progress.currentExplorationCheckpoint.helpIndex
+          trackedAnswerCount = progress.currentExplorationCheckpoint.pendingUserAnswersCount
+          hintSequenceNumber = 0
+          delayToShowNextHintAndSolution = -1
+        }.build()
+      }
+  }
+
+  /**
+   * Initializes the variables of [StateDeck]. If the [ExplorationCheckpoint] is of type default
+   * instance, the values of [StateDeck] are reset. Otherwise, the variables of [StateDeck] are
+   * re-initialized with the values created from the saved [ExplorationCheckpoint].
+   *
+   * This function expects explorationProgress.hintState to be initialized with the correct values,
+   * so it should only be called after the function [loadHintState] has executed.
+   */
+  private fun loadStateDeck(progress: ExplorationProgress, exploration: Exploration) {
+    if (progress.currentExplorationCheckpoint == ExplorationCheckpoint.getDefaultInstance()) {
+      progress.stateDeck.resetDeck(progress.stateGraph.getState(exploration.initStateName))
+    } else {
+      progress.stateDeck.resumeDeck(
+        createPendingTopStateFromCheckpoint(progress),
+        getPreviousStatesFromCheckpoint(progress),
+        progress.currentExplorationCheckpoint.pendingUserAnswersList,
+        progress.currentExplorationCheckpoint.stateIndex
+      )
+    }
+  }
```
This feels a bit like something that should be in the progress class itself since it's all about properly initializing the progress state. I suggest moving it over to keep this controller a bit more focused on its direct public API implementations.

MaskedCarrot replied:
moved to exploration progress

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
+    pendingStateHintList: List<Hint>,
+    helpIndex: HelpIndex
+  ): List<Hint> {
+    val updatedHintList: MutableList<Hint> = ArrayList()
+    when (helpIndex.indexTypeCase) {
+      HelpIndex.IndexTypeCase.HINT_INDEX -> {
+        pendingStateHintList.forEachIndexed { index, hint ->
+          if (index < helpIndex.hintIndex) {
+            // Mark all hints as visible and revealed which have an index less than that stored in
+            // the HintState.
+            updatedHintList.add(hint.toBuilder().setHintIsRevealed(true).build())
+          } else if (index == helpIndex.hintIndex) {
+            // Add the currently shown hint to the updated hint list and set hintIsRevealed
+            // depending upon if the current hint was revealed or not.
+            updatedHintList.add(
+              hint.toBuilder().setHintIsRevealed(helpIndex.isAllVisibleHelpRevealed).build()
+            )
+          } else {
+            // Add all the remaining hints that are not yet visible to the user to the updated hint
+            // list.
+            updatedHintList.add(
+              hint.toBuilder().setHintIsRevealed(false).build()
+            )
+          }
+        }
+      }
+      HelpIndex.IndexTypeCase.SHOW_SOLUTION, HelpIndex.IndexTypeCase.EVERYTHING_REVEALED -> {
+        // All the hints are visible and revealed if helpIndex.indexTypeCase is equal to
+        // SHOW_SOLUTION or EVERYTHING_REVEALED.
+        pendingStateHintList.forEach { hint ->
+          updatedHintList.add(hint.toBuilder().setHintIsRevealed(true).build())
+        }
+      }
+      else -> updatedHintList.addAll(pendingStateHintList)
+    }
+    return updatedHintList
```
```suggestion
    return when (helpIndex.indexTypeCase) {
      HelpIndex.IndexTypeCase.HINT_INDEX -> {
        pendingStateHintList.forEachIndexed { index, hint ->
          if (index < helpIndex.hintIndex) {
            // Mark all hints as visible and revealed which have an index less than that stored in
            // the HintState.
            listOf(hint.toBuilder().setHintIsRevealed(true).build())
          } else if (index == helpIndex.hintIndex) {
            // Add the currently shown hint to the updated hint list and set hintIsRevealed
            // depending upon if the current hint was revealed or not.
            listOf(hint.toBuilder().setHintIsRevealed(helpIndex.isAllVisibleHelpRevealed).build())
          } else {
            // Add all the remaining hints that are not yet visible to the user to the updated hint
            // list.
            listOf(hint.toBuilder().setHintIsRevealed(false).build())
          }
        }
      }
      HelpIndex.IndexTypeCase.SHOW_SOLUTION, HelpIndex.IndexTypeCase.EVERYTHING_REVEALED -> {
        // All the hints are visible and revealed if helpIndex.indexTypeCase is equal to
        // SHOW_SOLUTION or EVERYTHING_REVEALED.
        pendingStateHintList.forEach { hint ->
          listOf(hint.toBuilder().setHintIsRevealed(true).build())
        }
      }
      else -> pendingStateHintList
    }
```

Better to avoid mutable state. Ditto elsewhere.

MaskedCarrot replied:
changed to mapIndexed

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
+  /**
+   * Creates a list of completed states from the saved [ExplorationCheckpoint].
+   *
+   * @return [List] of [EphemeralState] containing all the states that were completed before the
```
```suggestion
   * @return [List] of [EphemeralState]s containing all the states that were completed before the
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
+  ): List<EphemeralState> {
+    val previousStates: MutableList<EphemeralState> = ArrayList()
+    progress.currentExplorationCheckpoint.completedStatesInCheckpointList
+      .forEachIndexed { index, state ->
```
Per the above suggestion, you could use ``mapIndexed`` here, instead, to avoid the mutable list.

MaskedCarrot replied:
changed to mapIndexed

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
+          helpIndex = progress.currentExplorationCheckpoint.helpIndex
+          trackedAnswerCount = progress.currentExplorationCheckpoint.pendingUserAnswersCount
+          hintSequenceNumber = 0
+          delayToShowNextHintAndSolution = -1
```
I think this means that the next hint will default to the first hint's initial triggering criteria vs. the "next hint" criteria (per the diagram in HintHandler's documentation). Is that expected from a UX/product perspective?

MaskedCarrot replied:
actually, it does not matter what we set the delay here because the delay is recomputed based on the ephemeral state before anything is visible to the user. I set it to -1 as a default value because elsewhere we are using -1 to indicate that no timer should be scheduled.

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. Reviewed everything per my codeowners except ExplorationProgressControllerTest. The mechanism overall seems good, but I suspect it may change a bit per the base PR.

Please also update the title of this PR & downstream ones to include the PR that they're subsequently blocked on & using as a base branch (to make it clear that the PR is blocked & not yet ready for merging).

MaskedCarrot committed:
moved timer to domain

MaskedCarrot committed:
fixed build and nits

MaskedCarrot committed:
merge base branch

MaskedCarrot committed:
moved functions to ExplorationProgress.kt

MaskedCarrot committed:
nit

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
added listener back to test exemptions

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgress.kt:
```diff
@@ -113,4 +119,144 @@ internal class ExplorationProgress {
+   * This function expects explorationProgress.hintState to be initialized with the correct values,
+   * so it should only be called after the function [loadHintState] has executed.
+   */
+  internal fun loadStateDeck(exploration: Exploration) {
```
Add a `@param` annotation here

MaskedCarrot replied:
done

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgress.kt:
```diff
@@ -113,4 +119,144 @@ internal class ExplorationProgress {
+   * state before the checkpoint was saved.
+   *
+   * @param pendingStateHintList the list of hint for the current pending state
+   * @param helpIndex the state of hints for the exploration which was generated using the saved
```
Add a `@return` annotation as well

MaskedCarrot replied:
done

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgress.kt:
```diff
@@ -113,4 +119,144 @@ internal class ExplorationProgress {
+   * revealed for the current state pending state before the checkpoint was saved.
+   *
+   * @param pendingTopState the pending state created from the checkpoint
+   * @param helpIndex the state of solution for the exploration which was generated using the saved
```
Add an `@return` annotation too

MaskedCarrot replied:
done

vinitamurthi commented on model/src/main/proto/exploration_checkpoint.proto:
```diff
@@ -30,24 +30,17 @@ message ExplorationCheckpoint {
-
   // The title of the exploration whose checkpoint is saved.
-  string exploration_title = 7;
+  string exploration_title = 5;
```
Avoid modifying existing proto fields. Instead you can deprecate fields 5 and 6 and keep 7 as is.
i.e. 
```
reserved 5,6;

// ....
string exploration_title = 7;

// ...
string exploration_version = 8;

.....
HelpIndex help_index = 10;
```

MaskedCarrot replied:
done

MaskedCarrot committed:
nit

MaskedCarrot committed:
updated proto

MaskedCarrot committed:
nit fixes and added more tests

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
removed unused fun

rt4914 reviewed (approved):
Approved for code-owner files.

BenHenning committed:
First round: make HintHandler independent.

This also brings HintHandler into an interface + factory pattern. This
isn't the final design since I think we can largely simplify the way
hints work; that'll be my next pass.

This breaks questions & HintHandlerTest; those will require further work
later.

vinitamurthi reviewed (approved):
LGTM for codeowner files

BenHenning committed:
Simplify HelpIndex in proto.

Move HelpIndex to PendingState to avoid the entire domain case of
handling CompletedState.

BenHenning committed:
Simplify hints & solutions.

This removes the per-Hint/Solution tracking & completely leans on
HelpIndex for proper hints & solution tracking both in the domain & UI
layers.

Fixes a bunch of other stuff, too, including
QuestionAssessmentProgressController tests.

BenHenning committed:
Clean up dead code paths & improve handler API.

yashraj-01 committed:
fixed test modules and lint

yashraj-01 committed:
renamed HintHandlerTest to HintHandlerImplTest

BenHenning committed:
Add tests for HintHandler.

This introduces some new explorations for making testing HintHandlerImpl
easier.

BenHenning committed:
Add remaining tests/exemptions for new files.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	app/src/main/java/org/oppia/android/app/application/ApplicationComponent.kt
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Post-merge fixes (including lint fixes).

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Post-merge maven_install fix.

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

MaskedCarrot committed:
Revert "Merge branch 'develop' into move-hint-handler-to-domain"

This reverts commit e2fea908a65af8c11b5ed135f59cec38e807f251, reversing
changes made to 6659858f0b85a0bf0c51f23af600769eac64b938.

BenHenning committed:
Post-merge Gradle-discovered fixes.

BenHenning committed:
Revert "Revert "Merge branch 'develop' into move-hint-handler-to-domain""

This reverts commit b1622c00d33f584ae727f0c9c4434ad773bc80a2.

BenHenning committed:
Merge branch 'attempt-to-fix-move-hint-handler-to-domain' into move-hint-handler-to-domain

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning commented:
Sorry, I don't have enough focus to review this in detail right now. However, I also suspect that this is going to change a bit after #3659 has my changes incorporated since I removed the non-HelpIndex pieces from ExplorationCheckpoint & changed a few other things. I'm certain that we no longer need to pass along the sequence number or other data, and I'm pretty sure the changes from #3659 will be sufficient for hint/solution restoration to "just work" when loading a checkpoint. @MaskedCarrot please update this PR with the changes from there once they are ready & then assign back.

BenHenning committed:
Additional post-merge fixes.

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

anandwana001 reviewed (approved):
LGTM for code owners file. 

Could you add a espresso test result screenshot for future reference fo this file - app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt

MaskedCarrot committed:
updated appraoch to resume hints to work with the new hint handler.

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -48,6 +48,15 @@ import org.oppia.android.app.model.State
 interface HintHandler {
 
+  /**
+   * Restores the local variables of hint handler to a specific point in the exploration.
```
```suggestion
   * Restores the hint handler to a previously cached state so that the user can continue where they
   * left off.
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -48,6 +48,15 @@ import org.oppia.android.app.model.State
+   * Restores the local variables of hint handler to a specific point in the exploration.
+   *
+   * @param trackedWrongAnswerCount the count of wrong answers saved in the checkpoint
+   * @param helpIndex the [HelpIndex] saved in the checkpoint.
```
```suggestion
   * @param helpIndex the cached state of hints/solution from the checkpoint
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -48,6 +48,15 @@ import org.oppia.android.app.model.State
+   *
+   * @param trackedWrongAnswerCount the count of wrong answers saved in the checkpoint
+   * @param helpIndex the [HelpIndex] saved in the checkpoint.
+   * @param hintCount the total number of hint available on the saved pending state
```
```suggestion
   * @param hintCount the total number of hints available in the restored pending state
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -48,6 +48,15 @@ import org.oppia.android.app.model.State
+   * @param helpIndex the [HelpIndex] saved in the checkpoint.
+   * @param hintCount the total number of hint available on the saved pending state
+   * */
+  fun restoreHintHandler(trackedWrongAnswerCount: Int, helpIndex: HelpIndex, hintCount: Int)
```
Please add HintHandler-specific tests for the new behaviors introduced in this method.

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandlerImpl.kt:
```diff
@@ -33,11 +35,55 @@ class HintHandlerImpl private constructor(
       pendingState = state
       hintMonitor.onHelpIndexChanged()
-      maybeScheduleShowHint(wrongAnswerCount = 0)
+      maybeScheduleShowHint(trackedWrongAnswerCount)
```
This is semantically a bit strange. We'd expect a new state to have no answers to continue with.

Maybe instead we should have an API for HintHandler that's resuming for a given state, e.g. something like:

``fun resumeHintsForSavedState(trackedWrongAnswerCount: Int, helpIndex: HelpIndex, state: State)``

where we call that instead of this method in the event that we're restoring things. I think that'll simplify stuff a bit since it also avoids needing to restore the HintHandler in ExplorationProgress (which seems a bit odd since that class otherwise doesn't care at all about HintHandler).

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgress.kt:
```diff
@@ -110,4 +114,67 @@ internal class ExplorationProgress {
+   * @param exploration the current exploration that is being played
+   * @param hintHandler the [HintHandler] that will be used to show help in the current exploration
+   */
+  internal fun resetOrResumeExploration(exploration: Exploration, hintHandler: HintHandler) {
```
I suggest instead just making this resume & doing the check at the callsite since it avoids needing to recheck the state in multiple places. You can also reduce the hedging in the documentation since things become much more crisp: it's just restoring state, no conditions.

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2235,154 +2338,797 @@ class ExplorationProgressControllerTest {
+      mockCurrentStateLiveDataObserver,
+      atLeastOnce()
+    ).onChanged(currentStateResultCaptor.capture())
+    assertThat(currentStateResultCaptor.value.isSuccess()).isTrue()
```
Here & elsewhere in the updated/new tests: what is this verification attempting to ensure? It seems unnecessary.

Edit: I realized this is needed for tests that then use the retrieved state so I revise this comment to only correspond to tests that don't use the current state.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2235,154 +2338,797 @@ class ExplorationProgressControllerTest {
+    assertThat(currentStateResultCaptor.value.isSuccess()).isTrue()
+    val currentState = currentStateResultCaptor.value.getOrThrow()
+    assertThat(currentState.checkpointState)
+      .isEqualTo(CheckpointState.CHECKPOINT_UNSAVED)
```
Can this fit on one line?

```suggestion
    assertThat(currentState.checkpointState).isEqualTo(CheckpointState.CHECKPOINT_UNSAVED)
```

Prefer to pack everything onto one line unless it needs to wrap. Ditto elsewhere.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2235,154 +2338,797 @@ class ExplorationProgressControllerTest {
+  }
+
+  @Test
+  fun testCheckpointing_saveCheckpoint_checkCheckpointStateIsCorrect() {
```
```suggestion
  fun testCheckpointing_saveCheckpoint_checkCheckpointStateIsSuccessful() {
```

We should be a bit more specific as to what 'correct' is when we can be.

Ditto for the other new tests.

MaskedCarrot replied:
updated test names

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2235,154 +2338,797 @@ class ExplorationProgressControllerTest {
+    assertThat(currentStateResultCaptor.value.isSuccess()).isTrue()
+    val currentState = currentStateResultCaptor.value.getOrThrow()
+    assertThat(currentState.checkpointState)
+      .isEqualTo(CheckpointState.CHECKPOINT_SAVED_DATABASE_EXCEEDED_LIMIT)
```
Why was this hit in this test? I'm not seeing anything in the arrangement that would result in the database being full.

MaskedCarrot replied:
I have changed the size limit allocated to the checkpoint database, for testing this controller the size is set to 150 Bytes. This value is set so that the checkpoint database exceeds the allocated size limit when the second checkpoint is saved for TEST_EXPLORATION_ID_2

BenHenning replied:
Ah. Can you add a single line comment explaining that here to help contextualize it?

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2235,154 +2338,797 @@ class ExplorationProgressControllerTest {
+    submitWrongAnswerForPrototypeState2()
+    // Reveal the hint, then submit another wrong answer to trigger the solution.
+    verifyOperationSucceeds(explorationProgressController.submitHintIsRevealed(hintIndex = 0))
+
+    testCoroutineDispatchers.advanceTimeBy(TimeUnit.SECONDS.toMillis(30))
+
+    endExploration()
```
```suggestion
    testCoroutineDispatchers.advanceTimeBy(TimeUnit.SECONDS.toMillis(30))
    endExploration()
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2235,154 +2338,797 @@ class ExplorationProgressControllerTest {
+      shouldSavePartialProgress = true,
+      explorationCheckpoint = retrieveExplorationCheckpoint(profileId, TEST_EXPLORATION_ID_2)
+    )
+    // Verify that the helpIndex.IndexTypeCase is equal EVERYTHING_IS_REVEALED because all available
```
```suggestion

    // Verify that the helpIndex.IndexTypeCase is equal EVERYTHING_IS_REVEALED because all available
```

For this & the above comment, we should make sure the arrange, act, and assert blocks are clearly spearated in the test.

Ditto for the other tests.

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2235,154 +2338,797 @@ class ExplorationProgressControllerTest {
+  }
 
+  @Test
+  fun testCheckpointing_revealedSolution_resumeExp_revealedSolIsVisibleOnPendingState() {
```
Overall, the new tests are really good.

I suggest also adding 1 test that completes a few states, resumes, completes a few more, then resumes again and finishes the exploration to demonstrate multiple start/stops & that a restored exploration can be successfully finished.

I also suggest adding a test that shows finishing an exploration results in no/an empty checkpoint being saved (unless this test already exists?).

MaskedCarrot replied:
added more tests.

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot! I really like how much the implementation has overall simplified.

Took a full pass and left comments--PTAL.

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
fixed merge

MaskedCarrot committed:
test and lint fix

MaskedCarrot committed:
updated

MaskedCarrot committed:
fixed tests

MaskedCarrot committed:
lint fix

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:
```diff
@@ -1468,6 +1472,19 @@ class StateFragmentTest {
   }
 
+  @Test
+  @RunOn(TestPlatform.ESPRESSO) // TODO(#1612): Enable for Robolectric.
```
Why doesn't this work on Robolectric?

MaskedCarrot replied:
This does not work on robolectric because we can't move past the drag and drop state of the exploration on robolectric yet. 

Also the logic for deleting progress is being tested in exploration activity test by the testing the altert dialog boxes.

BenHenning replied:
Ah, drag and drop is definitely an issue then. Thanks for clarifying.

And FWIW it should be possible to test AlertDialogs in Robolectric.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -23,21 +23,23 @@ import org.oppia.android.app.model.State
+  /**
+   * Starts watching for potential hints to be shown when the exploration is resumed. (e.g. if a
+   * user doesn't submit an answer after a certain amount of time). This is meant to only be called
+   * once when the the exploration is r resumed.
```
```suggestion
   * once when the the exploration is resumed.
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -23,21 +23,23 @@ import org.oppia.android.app.model.State
+   *
+   * @param trackedWrongAnswerCount the count of wrong answers saved in the checkpoint
+   * @param helpIndex the cached state of hints/solution from the checkpoint
+   * @param state the restored pending state.
```
```suggestion
   * @param state the restored pending state
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -23,21 +23,23 @@ import org.oppia.android.app.model.State
   fun startWatchingForHintsInNewState(state: State)
 
+  /**
+   * Starts watching for potential hints to be shown when the exploration is resumed. (e.g. if a
```
```suggestion
   * Starts watching for potential hints to be shown when the exploration is resumed (e.g. if a
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2982,11 +3017,49 @@ class ExplorationProgressControllerTest {
+  }
 
+  @Test
+  fun testCheckpointing_revealedHintIsVisible_resumeExp_Wait10Seconds_solutionIsNotVisible() {
```
```suggestion
  fun testCheckpointing_revealedHintIsVisible_resumeExp_wait10Seconds_solutionIsNotVisible() {
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -3009,6 +3083,45 @@ class ExplorationProgressControllerTest {
   }
 
+  @Test
+  fun testCheckpointing_revealedHintIsVisible_resumeExp_Wait30Seconds_solutionIsNotVisible() {
```
Ditto

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -3009,6 +3083,45 @@ class ExplorationProgressControllerTest {
     assertThat(currentState.isHintRevealed(0)).isTrue()
```
Should this instead be verifying that the solution isn't visible, and that the index type is not SHOW_SOLUTION?

MaskedCarrot replied:
added an assert statement to ensure the solution is not revealed. 
Also, checking that the index type is `last revealed hint index`  instead of verifying that it is not `show solution` also ensures that the index type case is not some other value (for example index type not set)

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -498,22 +500,34 @@ class ExplorationProgressController @Inject constructor(
     explorationProgress.stateDeck.getCurrentEphemeralState(computeCurrentHelpIndex())
 
-  private fun computeCurrentHelpIndex(): HelpIndex = hintHandler.getCurrentHelpIndex()
+  private fun computeCurrentHelpIndex(): HelpIndex =
+    hintHandler.getCurrentHelpIndex()
```
Suggest reverting since this can still fit on one line.

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandlerImpl.kt:
```diff
@@ -77,6 +79,53 @@ class HintHandlerImpl private constructor(
+          solutionIsAvailable = false
+          solutionIsRevealed = false
+        }
+        SHOW_SOLUTION -> {
+          // 1 is subtracted from the hint count because hints are indexed from 0.
+          lastRevealedHintIndex = state.interaction.hintCount - 1
+          latestAvailableHintIndex = state.interaction.hintCount - 1
+          solutionIsAvailable = true
+          solutionIsRevealed = false
+        }
+        EVERYTHING_REVEALED -> {
+          // 1 is subtracted from the hint count because hints are indexed from 0.
+          lastRevealedHintIndex = state.interaction.hintCount - 1
+          latestAvailableHintIndex = state.interaction.hintCount - 1
+          solutionIsAvailable = true
+          solutionIsRevealed = true
+        }
```
```suggestion
        SHOW_SOLUTION, EVERYTHING_REVEALED -> {
          // 1 is subtracted from the hint count because hints are indexed from 0.
          lastRevealedHintIndex = state.interaction.hintCount - 1
          latestAvailableHintIndex = state.interaction.hintCount - 1
          solutionIsAvailable = true
          solutionIsRevealed = helpIndex.indexTypeCase == EVERYTHING_REVEALED
        }
```

Simplification.

MaskedCarrot replied:
simplified

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandlerImpl.kt:
```diff
@@ -77,6 +79,53 @@ class HintHandlerImpl private constructor(
+  ) {
+    handlerLock.withLock {
+      when (helpIndex.indexTypeCase) {
+        NEXT_AVAILABLE_HINT_INDEX -> {
+          lastRevealedHintIndex = helpIndex.nextAvailableHintIndex - 1
+          latestAvailableHintIndex = helpIndex.nextAvailableHintIndex
+          solutionIsAvailable = false
+          solutionIsRevealed = false
+        }
+        LATEST_REVEALED_HINT_INDEX -> {
+          lastRevealedHintIndex = helpIndex.latestRevealedHintIndex
+          latestAvailableHintIndex = helpIndex.latestRevealedHintIndex
+          solutionIsAvailable = false
+          solutionIsRevealed = false
+        }
```
```suggestion
        NEXT_AVAILABLE_HINT_INDEX, LATEST_REVEALED_HINT_INDEX -> {
          lastRevealedHintIndex = if (helpIndex.indexTypeCase == NEXT_AVAILABLE_HINT_INDEX) {
            helpIndex.nextAvailableHintIndex - 1
          } else helpIndex.nextAvailableHintIndex
          latestAvailableHintIndex = helpIndex.nextAvailableHintIndex
          solutionIsAvailable = false
          solutionIsRevealed = false
        }
```

Simplification (may require directly importing NEXT_AVAILABLE_HINT_INDEX).

MaskedCarrot replied:
done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -1622,7 +1662,7 @@ class ExplorationActivityTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       TestExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
```suggestion
      NetworkConnectionDebugUtilModule::class
```

MaskedCarrot replied:
done while resolving conflicts

BenHenning replied:
Thanks for clarifying that! Sometimes it's hard to review merge commits.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintHandlerImplTest.kt:
```diff
@@ -199,6 +199,239 @@ class HintHandlerImplTest {
+  @Test
+  fun testResumeHints_stateWithHints_hintVisible_wait10Seconds_doesNotCallMonitorAgain() {
+    val state = expWithHintsAndSolution.getInitialState()
+
```
```suggestion
```

The ``val state`` bit is part of the arrangement block. Ditto for other similar tests.

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintHandlerImplTest.kt:
```diff
@@ -199,6 +199,239 @@ class HintHandlerImplTest {
+
+    waitFor30Seconds()
+
+    // Verify that the monitor is called again (since there's a solution now available).
+    verify(mockHintMonitor).onHelpIndexChanged()
```
For each of these new tests where the monitor is called, please also add corresponding tests that verify the current HelpIndex state.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintHandlerImplTest.kt:
```diff
@@ -199,6 +199,239 @@ class HintHandlerImplTest {
+  }
+
+  @Test
+  fun testResumeHints_stateWithHints_hintReveled_wait30Seconds_callsMonitorAgain() {
```
These are really nice tests to verify that restoration works as expected. Please also add the two other missing cases & related wait tests) for SHOW_SOLUTION and EVERYTHING_REVEALED.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2379,8 +2541,744 @@ class ExplorationProgressControllerTest {
+    playThroughPrototypeState10AndMoveToNextState()
+
+    // Verify that the last state is terminal.
+    verify(mockCurrentStateLiveDataObserver, atLeastOnce())
+      .onChanged(currentStateResultCaptor.capture())
+    assertThat(currentStateResultCaptor.value.isSuccess()).isTrue()
+    val currentState = currentStateResultCaptor.value.getOrThrow()
+    assertThat(currentState.stateTypeCase).isEqualTo(TERMINAL_STATE)
```
Can we also verify that the checkpoint is empty?

MaskedCarrot replied:
There will be no empty checkpoint saved here, the most recent checkpoint that is saved in this case is when the user comes to the terminal state (i.e. the pending state changes). After that when the user clicks on the "Return to Topic" button, the saved progress for this lesson will be deleted.

MaskedCarrot replied:
We are verifying that the progress is deleted successfully on clicking the "Return to topic" button in StateFragmentTest

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -2227,6 +2327,36 @@ class ExplorationProgressControllerTest {
+    submitWrongAnswerForPrototypeState2()
+    submitWrongAnswerForPrototypeState2()
+
+    verify(mockCurrentStateLiveDataObserver, atLeastOnce())
+      .onChanged(currentStateResultCaptor.capture())
+    assertThat(currentStateResultCaptor.value.isSuccess()).isTrue()
```
Copying from https://github.com/oppia/oppia-android/pull/3668#discussion_r691835241 since that comment seemed to break on GitHub somehow.

Here & elsewhere in the updated/new tests: what is this verification attempting to ensure? It seems unnecessary.

Edit: I realized this is needed for tests that then use the retrieved state so I revise this comment to only correspond to tests that don't use the current state.

MaskedCarrot replied:
done

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. Took another pass--I think the PR is looking quite good. Just had a variety of suggestions, nits, and some requests for improved testing. PTAL.

BenHenning commented:
Also @MaskedCarrot please sync with develop & resolve conflicts.

MaskedCarrot committed:
Merge branch 'develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
improved testing and nit fix

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandlerImpl.kt:
```diff
@@ -86,31 +86,23 @@ class HintHandlerImpl private constructor(
-          lastRevealedHintIndex = helpIndex.nextAvailableHintIndex - 1
-          latestAvailableHintIndex = helpIndex.nextAvailableHintIndex
+        NEXT_AVAILABLE_HINT_INDEX, LATEST_REVEALED_HINT_INDEX -> {
+          if (helpIndex.indexTypeCase == NEXT_AVAILABLE_HINT_INDEX) {
+            lastRevealedHintIndex = helpIndex.nextAvailableHintIndex - 1
+            latestAvailableHintIndex = helpIndex.nextAvailableHintIndex
+          } else {
+            lastRevealedHintIndex = helpIndex.latestRevealedHintIndex
+            latestAvailableHintIndex = helpIndex.latestRevealedHintIndex
+          }
```
Ah, I misread the code before. Honestly, it's probably more readable the other way since we can't simplify it like the solution cases.

MaskedCarrot replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintHandlerImplTest.kt:
```diff
@@ -414,9 +424,26 @@ class HintHandlerImplTest {
+  }
+
+  @Test
+  fun testResumeHints_stateWithHints_firstHintReveled_wait30Seconds_helpIndexHasSolution() {
```
```suggestion
  fun testResumeHints_stateWithOneHint_hintRevealed_wait30Seconds_helpIndexHasSolution() {
```

Otherwise it's slightly confusing that it's jumping straight to the solution.

Also, typo in 'revealed' here & elsewhere. Please fix.

MaskedCarrot replied:
done

BenHenning reviewed (commented):
Thanks @MaskedCarrot. LGTM + 2 nits.

Also, PlatformParameterSyncUpWorkerTest seems like a potential unrelated flake (/cc @ARJUPTA). I'm following up with Arjun, but fixing the above should also retrigger the CI workflow.

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
fix failing test

BenHenning reviewed (approved):
Thanks @MaskedCarrot. LGTM!
